### ISSUE TITLE
Squash does not work correctly for AbstractGraph

### ISSUE THREAD

**Issue by simonschoelly** [JuliaGraphs/Graphs.jl#88]
The way the `squash` is currently implemented is by either calling `SimpleGraph{T}(g)` or `SimpleDiGraph{T}(g)` for some integer type `T`. This fails if `g` is a subtype of `AbstractGraph` for which these functions are not implemented. There are multiple ways to solve this:

1. Restrict `squash` to `SimpleGraph` and `SimpleDiGraph` and leave the implementation of that function for other graph types open.
2. Ensure that the constructors `SimpleGraph{T}(g)` or `SimpleDiGraph{T}(g)` work for for any kind of `g::AbstractGraph`.
3. Require that subtypes of any kind of graph  type `SomeGraph <: AbstractGraph` implement a constructor `SomeGraph{T}(g)` and adjust `squash` accordingly.

In my opinion, 1. is the way to go. 2. could lead to some information (such as metadata) being thrown away by `squash` which is probably not what the user wants. 3. would make the interface more complicated and would make existing packages not correspond to the interface anymore.

**Comment by etiennedeg** [comment]
I also think 1. is the way to go. It should nonetheless be possible to implement a conversion from `AbstractGraph` to `SimpleGraph` by collecting neighbors.

**Comment by simonschoelly** [comment]
Yes, not with the squash function though. One way to do that is to implement functions
```julia
SimpleGraph(::AbstractGraph)
SimpleGraph{T}(::AbstractGraph)
```
 or with a different function name, to keep the number of constructs a bit low. We would also need to figure out what to do in cases when the argument is a directed graph. Currently there is already a constructor `SimpleGraph(::SimpleDiGraph)` which adds an undirected edge, if there is one edge in either direction in the directed graph.

**Comment by simonschoelly** [comment]
Solved by #93
